rbd: implement live migration functions

omidyoosefi opened the pull request:
## Summary

This PR adds support for `rbd migration` apis allowing users to migrate images between pools, in the same pool or from an external stream. https://docs.ceph.com/en/pacific/rbd/rbd-live-migration/)

Signed-off-by: Omid Yoosefi [omidyoosefi@ibm.com](mailto:omidyoosefi@ibm.com)


## Checklist
- [x] Added tests for features and functional changes
- [x] Public functions and types are documented
- [x] Standard formatting is applied to Go code
- [x] Is this a new API? Added a new file that begins with `//go:build ceph_preview`

phlogistonjohn commented:
Hi @omidyoosefi thanks for the new contribution! I've enabled the CI for your PR.

I noticed that you left the "Is this a new API?" checklist item unchecked and the new files do not have a `ceph_preview` tag. Our definition of a new API means any new functions that are exposed to consumer code - so this would count. Could you please add the tag to the files?  If you have questions about how/why - feel free to ask.

One additional word of caution is that most of the go-ceph maintainers work on the project as part of our jobs and many of us are taking a break before the end of the year. For example, my vacation begins Friday and I only return to work in Jan.  This means that reviews of your contribution may be a bit slower than usual. That doesn't mean we're ignoring your PR. I'll try to provide some concrete feedback later today and probably will spend a little time on upstream tasks in go-ceph during my break. Do note that an API change does need two approving reviews before it gets merged.

omidyoosefi commented:
@phlogistonjohn Hi John, appreciate you taking a look at this so quickly! I've made the changes and repushed.

I understand that this could take some time to get merged but wanted to get the PR opened and some eyes on it for feedback.

phlogistonjohn commented on rbd/migration.go:
```diff
@@ -0,0 +1,240 @@
+	return getError(ret)
+}
+
+// MigrationExecute executes a migration starting
+// the full migration of data into image from source
```
Style nitpick. The first line of a doc comment should be a full english sentence that starts with the name of the item being documented and ending with a period.

omidyoosefi replied:
Changed

phlogistonjohn commented on rbd/migration.go:
```diff
@@ -0,0 +1,240 @@
+	return getError(ret)
+}
+
+// MigrationImageState denotes the current migration status of a given image
+type MigrationImageState int
+
+const (
+	// MigrationImageUnknown is the representation of
+	// RBD_IMAGE_MIGRATION_STATE_UNKNOWN from librbd.
+	MigrationImageUnknown = MigrationImageState(C.RBD_IMAGE_MIGRATION_STATE_UNKNOWN)
+	// MigrationImageError is the representation of
+	// RBD_IMAGE_MIGRATION_STATE_ERROR from librbd.
+	MigrationImageError = MigrationImageState(C.RBD_IMAGE_MIGRATION_STATE_ERROR)
+	// MigrationImagePreparing is the representation of
+	// RBD_IMAGE_MIGRATION_STATE_PREPARING from librbd.
+	MigrationImagePreparing = MigrationImageState(C.RBD_IMAGE_MIGRATION_STATE_PREPARING)
+	// MigrationImagePrepared is the representation of
+	// RBD_IMAGE_MIGRATION_STATE_PREPARED from librbd.
+	MigrationImagePrepared = MigrationImageState(C.RBD_IMAGE_MIGRATION_STATE_PREPARED)
+	// MigrationImageExecuting is the representation of
+	// RBD_IMAGE_MIGRATION_STATE_EXECUTING from librbd.
+	MigrationImageExecuting = MigrationImageState(C.RBD_IMAGE_MIGRATION_STATE_EXECUTING)
+	// MigrationImageExecuted is the representation of
+	// RBD_IMAGE_MIGRATION_STATE_EXECUTED from librbd.
+	MigrationImageExecuted = MigrationImageState(C.RBD_IMAGE_MIGRATION_STATE_EXECUTED)
+	// MigrationImageAborting is the representation of
+	// RBD_IMAGE_MIGRATION_STATE_ABORTING from librbd.
+	MigrationImageAborting = MigrationImageState(C.RBD_IMAGE_MIGRATION_STATE_ABORTING)
+)
```
Style nitpick: I generally like to see constants defined near the top of the file, before we start defining structures (unless there are many structures in the with with many different constants in the same file).

omidyoosefi replied:
Done

phlogistonjohn commented on rbd/migration.go:
```diff
@@ -0,0 +1,240 @@
+	}()
+
+	var status C.rbd_image_migration_status_t
+	ret := C.rbd_migration_status(
```
I think that, on success, this function call will allocate memory for the values stored in the                                       rbd_image_migration_status_t. The librbd header provides a function `rbd_migration_status_cleanup` to free this memory. We need to call this function after converting the C types to Go types in order to free the memory that the `rbd_migration_status` call allocated.

phlogistonjohn replied:
FWIW, you can double check the ceph code to see if `rbd_migration_status_cleanup` is a noop when `rbd_image_migration_status_t*` or the fields are null and if so implementing the cleanup can be done with a simple defer.

phlogistonjohn commented on rbd/migration_test.go:
```diff
@@ -0,0 +1,174 @@
+
+	// get status
+	status, err := MigrationStatus(ioctx, opts.DestImage)
+	assert.NoError(t, err)
```
This assertion appears to be failing in the CI.

phlogistonjohn commented on rbd/migration_test.go:
```diff
@@ -0,0 +1,174 @@
+	// get status
+	status, err := MigrationStatus(ioctx, opts.DestImage)
+	assert.NoError(t, err)
+	assert.Equal(t, status.State, MigrationImagePrepared)
```
This line is panicking in the CI, because (AFAICT) status is nil.

phlogistonjohn commented on rbd/migration_test.go:
```diff
@@ -0,0 +1,174 @@
+import (
+	"testing"
+
+	"github.com/ceph/go-ceph/rados"
```
Style nit: put a blank line between the (local) go-ceph imports and other imports.

omidyoosefi replied:
Done.

phlogistonjohn commented on rbd/migration_test.go:
```diff
@@ -0,0 +1,174 @@
+	// call MigrationPrepareImport
+	var opts MigrationOptions
+	opts.DestImage = GetUUID()
+	opts.SourceSpec = `{"type": "native", "snap_name": "snap1", "pool_name":"` + pool + `", "image_name": "` + name + `"}}`
```
Is the JSON structure for a Source Spec documented by Ceph? If so it may be useful to prove a function that generates the JSON string rather than just "hand assembling" it like this test does. If it's more ad-hoc I won't object to the test building the JSON string this way... but I think it's worth considering doing it in a more structured manner.

omidyoosefi replied:
The json spec seems very specific to the user's needs but I could try implementing helper functions for each. I'm thinking something like creating multiple functions such as `CreateSourceSpecS3`, `CreateSourceSpecHttp`, `CreateSourceSpecFile` and `CreateSourceSpecNative`? 


ref: https://docs.ceph.com/en/pacific/rbd/rbd-live-migration/#formats

phlogistonjohn commented on rbd/migration_test.go:
```diff
@@ -0,0 +1,174 @@
+	// setup image
+	name := createAndWriteDataToImage(t, ioctx)
+
+	// call MigrationPrepare
```
These sorts of comments can be removed. They're great while you're still writing your code as a placeholder but they add little to no value to a reader in the future.

omidyoosefi replied:
Done

phlogistonjohn commented on rbd/migration_test.go:
```diff
@@ -0,0 +1,174 @@
+	_, err = OpenImage(ioctx, opts.DestImage, NoSnapshot)
+	assert.NoError(t, err)
+
+	// original image is moved to trash as part of migration prepare
```
This comment is fine as it explains the why of what the code will be doing.

phlogistonjohn reviewed (changes requested):
First off - thanks a ton for the contribution. Most of the code looks nice and clean and this appears to be a useful feature to have.

I've provided some feedback including some style things and some more important structural notes. The most important thing we need to handle is the cleanup in the MigrationStatus function. We need to resolve that or we will leak memory.

Currently, the code is not passing the CI tests. I poked at it a little bit and I'm unclear if the destination image should exist before MigrationExecute has been called. Either that or something along those lines is triggering the initial set of test failures. On top of that there are panics due to the status function returning a nil status object.

Unfortunately, since you are a first time contributor I need to keep pressing a button in the github UI every time you submit new code for it to be run by the CI. In order to iterate more quickly you may want to run tests locally on a development system. If you're not familiar with how our tests can be run here's an example:
```
make test-containers-test ENTRYPOINT_ARGS="--test-pkg=rbd  --test-run=TestMig"  USE_CACHE=1
```
IIRC all that's required for your local system is docker or podman. This will run our full test infrastructure but only execute your test cases (ones starting with TestMig).  When you're all done testing or otherwise need to reset the ceph containers run `make test-containers-clean` Theres a bit more in our docs/ directory and don't hesitate to ask questions.

phlogistonjohn commented:
> Edit: I see that it requires approval for CI to run again, is that cause I force pushed?

Force pushing to your development branch is totally fine and generally how go-ceph expects PRs to be updated. As I alluded to in my other comment github has a policy of making every new set of changes (force pushed or otherwise) require a maintainer approve it. IIRC the policy came about because Github/Microsoft observed some PRs being used as away to abuse running code in azure for free.  While this is obviously an imperfect method to keep bad stuff from running on their infra I guess they think it's enough of a speedbump.

It'll keep happening on this PR.  If you push and you don't see the CI run for a day or so feel free to leave a comment "AT"ing me as a reminder to press that button.

omidyoosefi commented:
@phlogistonjohn thank you for the feedback. I'm making the changes right now. One question:
```
The most important thing we need to handle is the cleanup in the MigrationStatus function. We need to resolve that or we will leak memory.
```
Can you please elaborate or if possible show an example of how the cleanup is supposed to be done? I used some of the other APIs(`mirror_nautilus` to be exact) as an example.

omidyoosefi commented:
Btw I'm running into a weird issue with the tests...the failures seem intermittent and having trouble to really understand the reason cause there are no code changes between this sample of 3 consecutive runs.

- This run only one failed
```
*** running: go test -v -count=1 -tags=pacific,ceph_preview -run TestMig -covermode=count -coverprofile=rbd.cover.out -coverpkg=github.com/ceph/go-ceph/rbd ./rbd
=== RUN   TestMigration
    migration_test.go:57:
        	Error Trace:	/go/src/github.com/ceph/go-ceph/rbd/migration_test.go:57
        	Error:      	Received unexpected error:
        	            	RBD image not found
        	Test:       	TestMigration
--- FAIL: TestMigration (5.59s)
=== RUN   TestMigrationPrepareImport
--- PASS: TestMigrationPrepareImport (7.03s)
=== RUN   TestMigrationAbort
--- PASS: TestMigrationAbort (8.10s)
FAIL
```

- This run two failed
```
*** running: go test -v -count=1 -tags=pacific,ceph_preview -run TestMig -covermode=count -coverprofile=rbd.cover.out -coverpkg=github.com/ceph/go-ceph/rbd ./rbd
=== RUN   TestMigration
    migration_test.go:57:
        	Error Trace:	/go/src/github.com/ceph/go-ceph/rbd/migration_test.go:57
        	Error:      	Received unexpected error:
        	            	RBD image not found
        	Test:       	TestMigration
--- FAIL: TestMigration (6.13s)
=== RUN   TestMigrationPrepareImport
    migration_test.go:104:
        	Error Trace:	/go/src/github.com/ceph/go-ceph/rbd/migration_test.go:104
        	Error:      	Received unexpected error:
        	            	RBD image not found
        	Test:       	TestMigrationPrepareImport
--- FAIL: TestMigrationPrepareImport (6.25s)
=== RUN   TestMigrationAbort
--- PASS: TestMigrationAbort (7.59s)
FAIL
```

- This run 2 failed 
```
*** running: go test -v -count=1 -tags=pacific,ceph_preview -run TestMig -covermode=count -coverprofile=rbd.cover.out -coverpkg=github.com/ceph/go-ceph/rbd ./rbd
=== RUN   TestMigration
    migration_test.go:57:
        	Error Trace:	/go/src/github.com/ceph/go-ceph/rbd/migration_test.go:57
        	Error:      	Received unexpected error:
        	            	RBD image not found
        	Test:       	TestMigration
--- FAIL: TestMigration (5.64s)
=== RUN   TestMigrationPrepareImport
--- PASS: TestMigrationPrepareImport (7.05s)
=== RUN   TestMigrationAbort
    migration_test.go:137:
        	Error Trace:	/go/src/github.com/ceph/go-ceph/rbd/migration_test.go:137
        	Error:      	Received unexpected error:
        	            	RBD image not found
        	Test:       	TestMigrationAbort
--- FAIL: TestMigrationAbort (5.73s)
FAIL
```



I'm trying to chase down the issue but not really sure why it's intermittent failures, any guidance is appreciated.

omidyoosefi commented on rbd/migration.go:
```diff
@@ -0,0 +1,240 @@
+// 							  const char *image_name,
+// 							  rados_ioctx_t dest_ioctx,
+// 							  const char *dest_image_name,
+// 							  rbd_image_options_t opts);
```
Done

omidyoosefi commented on rbd/migration.go:
```diff
@@ -0,0 +1,240 @@
+// #undef _GNU_SOURCE
+// #include <errno.h>
+// #include <stdlib.h>
+// #include <rados/librados.h>
```
Done!

omidyoosefi commented:
Latest push addresses some of the comments. Still trying to figure out intermittent test failures as well as how to cleanly implement helper functions for SourceSpec JSON string generation.

phlogistonjohn commented:
The intermittent test failures may be due to race conditions that occur due to what may be asynchronous tasks being started by the ceph libraries. I have three suggestions, one for each test, that may resolve the intermittent test failures. I ran the tests 100 times in a row with these changes and they passed. However, I had a bunch of debugging Printfs and ceph logging enabled. I'm now running another loop of 100 in order to verify the tests continue to pass with the debugging and printfs removed.  I'll share the exact diffs *if* this works as I hope it will.

phlogistonjohn commented:
Unfortunately, it failed again - specifically I can get the TestMigration and TestMigrationPrepareImport working reliably with the status calls wrapped in a [testify.Eventually](https://pkg.go.dev/github.com/stretchr/testify/assert#Eventually).
However, I have struggled to make TestMigrationAbort reliable. I'm starting to wonder if there's an issue in the ceph libs. I'll try to capture any logging related to the failure but if the logging itself is changing the likelihood of the failure we may want to start a conversation with the librbd maintainers.  

I had hoped that closing the ioctx and opening a new one to call abort might flush any state preventing the abort call from succeeding, but it seems that about roughly 1/10 times (without ceph logging enabled) it fails with `RBD image not found`.

My WIP diff is below. I'm sharing it because I'm not sure when I'll be able to spend more time on this oever the next few days. So I'm sharing this now even though it doesn't fully work, in case you want to continue experimenting from this direction. 
```diff
diff --git a/rbd/migration_test.go b/rbd/migration_test.go
index df44079..0d69cbf 100644
--- a/rbd/migration_test.go
+++ b/rbd/migration_test.go
@@ -5,12 +5,33 @@ package rbd
 
 import (
 	"testing"
+	"time"
 
 	"github.com/ceph/go-ceph/rados"
+
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 )
@@ -53,15 +74,20 @@ func TestMigration(t *testing.T) {
 	err = MigrationPrepare(ioctx, ioctx, name, destImage, NewRbdImageOptions())
 	require.NoError(t, err)
 
-	status, err := MigrationStatus(ioctx, destImage)
-	require.NoError(t, err)
+	var status *MigrationImageInfo
+	require.Eventually(t, func() bool {
+		status, err = MigrationStatus(ioctx, destImage)
+		return err == nil
+	}, 5*time.Second, 10*time.Millisecond)
 	assert.Equal(t, status.State, MigrationImagePrepared)
 
 	err = MigrationExecute(ioctx, destImage)
 	require.NoError(t, err)
 
-	status, err = MigrationStatus(ioctx, destImage)
-	require.NoError(t, err)
+	require.Eventually(t, func() bool {
+		status, err = MigrationStatus(ioctx, destImage)
+		return err == nil
+	}, 5*time.Second, 10*time.Millisecond)
 	assert.Equal(t, status.State, MigrationImageExecuted)
 
 	err = MigrationCommit(ioctx, destImage)
@@ -100,8 +126,11 @@ func TestMigrationPrepareImport(t *testing.T) {
 	err = MigrationPrepareImport(ioctx, sourceSpec, destImage, NewRbdImageOptions())
 	require.NoError(t, err)
 
-	status, err := MigrationStatus(ioctx, destImage)
-	require.NoError(t, err)
+	var status *MigrationImageInfo
+	require.Eventually(t, func() bool {
+		status, err = MigrationStatus(ioctx, destImage)
+		return err == nil
+	}, 5*time.Second, 10*time.Millisecond)
 	assert.Equal(t, status.State, MigrationImagePrepared)
 
 	img, err := OpenImage(ioctx, destImage, NoSnapshot)
@@ -133,6 +162,10 @@ func TestMigrationAbort(t *testing.T) {
 	err = MigrationPrepare(ioctx, ioctx, name, destImage, NewRbdImageOptions())
 	require.NoError(t, err)
 
+	ioctx.Destroy()
+	ioctx, err = conn.OpenIOContext(pool)
+	require.NoError(t, err)
+
 	err = MigrationAbort(ioctx, destImage)
 	require.NoError(t, err)
 
```

omidyoosefi commented:
> ```diff
> +	require.Eventually(t, func() bool {
> +		status, err = MigrationStatus(ioctx, destImage)
> +		return err == nil
> +	}, 5*time.Second, 10*time.Millisecond)
> ```

Oh wow awesome, thank you!! It seems like there could definitely be a timing issue from the moment a MigrationPrepare is issued till the MigrationStatus can actually fetch the status causing the intermittent issue! I made the same change for waiting for the prepared status before the abort and here's a sample of 100 tries, 98 passed.

```
Behroozs-MacBook-Pro-2:go-ceph omid$ for i in $(seq 1 100); do res=$(make test-containers-test ENTRYPOINT_ARGS="--test-pkg=rbd  --test-run=TestMig"  USE_CACHE=1); echo $?;  done
0
0
0
0
0
0
0
0
0
0
0
make: *** [test-containers-test] Error 1
2
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
make: *** [test-containers-test] Error 1
2
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
```

omidyoosefi commented:
Latest push had 100% success rate on tests in a loop of 100 tries.

omidyoosefi commented:
Getting this error now, any guidance? 
```
Run if [ -f _results/implements.json ]; then
error: 1[8](https://github.com/ceph/go-ceph/actions/runs/3751624650/jobs/6384117351#step:6:9) problems detected
not tracked (preview): rbd MigrationPrepare
not tracked (preview): rbd MigrationPrepareImport
not tracked (preview): rbd MigrationExecute
not tracked (preview): rbd MigrationCommit
not tracked (preview): rbd MigrationAbort
not tracked (preview): rbd MigrationStatus
no added_in_version set: rbd MigrationPrepare
no expected_stable_version set: rbd MigrationPrepare
no added_in_version set: rbd MigrationPrepareImport
no expected_stable_version set: rbd MigrationPrepareImport
no added_in_version set: rbd MigrationExecute
no expected_stable_version set: rbd MigrationExecute
no added_in_version set: rbd MigrationCommit
no expected_stable_version set: rbd MigrationCommit
no added_in_version set: rbd MigrationAbort
no expected_stable_version set: rbd MigrationAbort
no added_in_version set: rbd MigrationStatus
no expected_stable_version set: rbd MigrationStatus
```

phlogistonjohn commented:
It looks to me like we have not resolved the reliability issue with the existing tests. Some of the CI runs are failing in the new tests and when I execute locally the new tests fail about 1/20 times (approximately).
I think the best bet is to contact the maintainers of this part of ceph rdb after the New Year holiday. 

As for the recent question: this is part of our API stability process. See [here](https://github.com/ceph/go-ceph/blob/master/docs/api-stability.md) and [here](https://github.com/ceph/go-ceph/blob/master/docs/development.md#api-status). 
I expect to have some more feedback on the APIs, specifically I noticed something about the argument order on one of the functions. Just be aware if you update the api status json file as documented and I, or another reviewer, asks for changes you may have to regenerate the JSON. It's up to you if you want to keep having to update that file or do it once after you've gotten more feedback.

As as side note, as the holidays are immanent I'll be even less online than I was this week. Thanks again for your contribution and I look forward to resuming working with you on this PR soon!

phlogistonjohn commented:
@trociny based on my search of the ceph sources I think you added the bulk of the migration rbd functions. If I'm correct, and you still work on this area of rbd I'd like to ask a few questions related to the APIs and the issues we're seeing with them in our test suite. If you don't work on this RBD feature can you point me at someone who could help? Thanks!

idryomov commented:
`debug ms = 1` + `debug rbd = 20` is my usual first choice.  Is that the one that makes the failures disappear?

Can you identify a test scenario where these spurious failures are most reproducible and describe it in detail?  Then I'd suggest to try to repeat the same scenario with C/C++ API and/or via CLI.

What do you refer to by "fail with `-ENOENT` much more than I would expect".  Do you actually expect such errors in your tests?

trociny commented:
Just FYI, we have a bunch of  C migration API tests, including abort/status calls [1] and we don't observe the issue like you describe.

[1] https://github.com/ceph/ceph/blob/main/src/test/librbd/test_Migration.cc#L584

phlogistonjohn commented:
I was using a big hammer and turning a lot more debugging up to 20. Either I'm getting more lucky today or your suggestions are better tuned to capture the types of errors we see. Here's a log file capturing the retry loop around MigrationStatus [timing out and failing](http://asynchrono.us/files/x-2023-01-06-MigrationStatusTimesout.txt.xz).

Here's snippets of the full log:
```2023-01-06T15:45:58.004+0000 7efbacff9700 10 librbd::ImageCtx: 0x7efbe800ba80 ~ImageCtx
2023-01-06T15:45:58.004+0000 7efbacff9700 20 librbd::AsioEngine: 0x7efbe800b870 ~AsioEngine: 
2023-01-06T15:45:58.004+0000 7efbacff9700 20 librbd::asio::ContextWQ: 0x7efbe8007930 ~ContextWQ: 
2023-01-06T15:45:58.004+0000 7efbacff9700 20 librbd::asio::ContextWQ: 0x7efbe8007930 drain: 2023-01-06T15:45:58.004+0000 7efbf3e9f700 10 librbd::Migration: open_images: Source image is not found. Trying trash
2023-01-06T15:45:58.004+0000 7efbf3e9f700 20 librbd::api::Trash: list: list 0x7efbf3e9b710
2023-01-06T15:45:58.004+0000 7efbf3e9f700 20 librbd::api::Trash: list_trash_image_specs: list_trash_image_specs 0x7efbf3e9b710
2023-01-06T15:45:58.004+0000 7efbf3e9f700  1 -- 10.89.0.25:0/2330621567 --> [v2:10.89.0.23:6800/144,v1:10.89.0.23:6801/144] -- osd_op(unknown.0.0:1589 11.0 11:c4c92e5a:::rbd_trash:head [call rbd.trash_list in=25b] snapc 0=[] ondisk+read+known_if_redirected+supports_pool_eio e44) v8 -- 0x7efbe8019850 con 0x1f3a520
2023-01-06T15:45:58.004+0000 7efbd2ffd700  1 -- 10.89.0.25:0/2330621567 <== osd.0 v2:10.89.0.23:6800/144 1602 ==== osd_op_reply(1589 rbd_trash [call] v0'0 uv0 ondisk = -2 ((2) No such file or directory)) v8 ==== 153+0+0 (crc 0 0 0) 0x7efbb4026040 con 0x1f3a520
2023-01-06T15:45:58.004+0000 7efbf3e9f700 -1 librbd::Migration: open_images: failed to determine image id: (2) No such file or directory
2023-01-06T15:45:58.011+0000 7efbf957f700 10 librbd::Migration: status: d1c0af05-d266-49b1-b47d-313ee957c85b/e57b985f-70e6-47f5-99ed-a52a97998e78
2023-01-06T15:45:58.011+0000 7efbf957f700 10 librbd::Migration: open_images: trying to open image by name d1c0af05-d266-49b1-b47d-313ee957c85b/e57b985f-70e6-47f5-99ed-a52a97998e78
2023-01-06T15:45:58.013+0000 7efbf957f700 20 librbd::asio::ContextWQ: 0x7efbec1a4490 ContextWQ: 
```
...
```
2023-01-06T15:45:58.014+0000 7efbcbfff700 10 librbd::image::CloseRequest: 0x1f85600 handle_shut_down_object_dispatcher: r=0
2023-01-06T15:45:58.014+0000 7efbcbfff700 10 librbd::image::CloseRequest: 0x1f85600 send_flush_op_work_queue
2023-01-06T15:45:58.014+0000 7efbcbfff700 10 librbd::image::CloseRequest: 0x1f85600 handle_flush_op_work_queue: r=0
2023-01-06T15:45:58.014+0000 7efbcbfff700 10 librbd::image::OpenRequest: handle_close_image: r=0
2023-01-06T15:45:58.014+0000 7efbcbfff700 10 librbd::ImageState: 0x7efbec16c6e0 0x7efbec16c6e0 handle_open: r=-2
2023-01-06T15:45:58.014+0000 7efbacff9700 10 librbd::ImageCtx: 0x7efbec153c10 ~ImageCtx
2023-01-06T15:45:58.014+0000 7efbacff9700 20 librbd::AsioEngine: 0x7efbec17bb10 ~AsioEngine: 
2023-01-06T15:45:58.014+0000 7efbacff9700 20 librbd::asio::ContextWQ: 0x7efbec1a4490 ~ContextWQ: 
2023-01-06T15:45:58.014+0000 7efbacff9700 20 librbd::asio::ContextWQ: 0x7efbec1a4490 drain: 
2023-01-06T15:45:58.014+0000 7efbf957f700 10 librbd::Migration: open_images: Source image is not found. Trying trash
2023-01-06T15:45:58.014+0000 7efbf957f700 20 librbd::api::Trash: list: list 0x7efbf957b710
2023-01-06T15:45:58.014+0000 7efbf957f700 20 librbd::api::Trash: list_trash_image_specs: list_trash_image_specs 0x7efbf957b710
2023-01-06T15:45:58.014+0000 7efbf957f700  1 -- 10.89.0.25:0/2330621567 --> [v2:10.89.0.23:6800/144,v1:10.89.0.23:6801/144] -- osd_op(unknown.0.0:1592 11.0 11:c4c92e5a:::rbd_trash:head [call rbd.trash_list in=25b] snapc 0=[] ondisk+read+known_if_redirected+supports_pool_eio e44) v8 -- 0x7efbec1fe0d0 con 0x1f3a520
2023-01-06T15:45:58.014+0000 7efbd2ffd700  1 -- 10.89.0.25:0/2330621567 <== osd.0 v2:10.89.0.23:6800/144 1605 ==== osd_op_reply(1592 rbd_trash [call] v0'0 uv0 ondisk = -2 ((2) No such file or directory)) v8 ==== 153+0+0 (crc 0 0 0) 0x7efbb4026040 con 0x1f3a520
2023-01-06T15:45:58.014+0000 7efbf957f700 -1 librbd::Migration: open_images: failed to determine image id: (2) No such file or directory
    migration_test.go:108: 
        »»»»Error Trace:»»»»/go/src/github.com/ceph/go-ceph/rbd/migration_test.go:108
        »»»»Error:      »»»»Condition never satisfied
        »»»»Test:       »»»»TestMigrationPrepareImport
--- FAIL: TestMigrationPrepareImport (11.16s)

```

I can see that it calls into open_images in ceph's src/librbd/api/Migration.cc, but it appears to fail to resolve the image name. However, it's not clear to me why as we're not "doing anything" to the images after running the prepare call.

phlogistonjohn commented:
I ran it a few more times and was able to reproduce the situation where the MigrationAbort call fails, after MigrationStatus has apparently succeeded. [Full log](http://asynchrono.us/files/x-2023-01-06-MigrationAbortFailsAfterMigrationStatusOK.txt.xz)

Snippets:
```
2023-01-06T16:08:57.857+0000 7fcd3d7fa700  1 -- 10.89.0.28:0/968080585 <== osd.0 v2:10.89.0.26:6800/144 253 ==== osd_op_reply(232 rbd_header.10bda7b0bbc4 [watch watch cookie 140518713929744] v50'49 uv40 ondisk = 0) v8 ==== 167+0+0 (crc 0 0 0) 0x7fcd1403d020 con 0x195b050
2023-01-06T16:08:57.857+0000 7fcd077fe700 10 librbd::Watcher: 0x7fcd10067640 handle_register_watch: r=0
2023-01-06T16:08:57.857+0000 7fcd077fe700 10 librbd::image::OpenRequest: 0x7fcd60099a30 handle_register_watch: r=0
2023-01-06T16:08:57.857+0000 7fcd077fe700  5 librbd::io::SimpleSchedulerObjectDispatch: 0x7fcd582455d0 SimpleSchedulerObjectDispatch: ictx=0x7fcd60051540
2023-01-06T16:08:57.857+0000 7fcd077fe700  5 librbd::io::SimpleSchedulerObjectDispatch: 0x7fcd582455d0 init: 
2023-01-06T16:08:57.857+0000 7fcd077fe700  5 librbd::io::Dispatcher: 0x7fcd600992d0 register_dispatch: dispatch_layer=5
2023-01-06T16:08:57.857+0000 7fcd077fe700 10 librbd::ImageState: 0x7fcd60099250 0x7fcd60099250 handle_open: r=0   
2023-01-06T16:08:57.857+0000 7fcd6d586700  1 -- 10.89.0.28:0/968080585 --> [v2:10.89.0.26:6800/144,v1:10.89.0.26:6801/144] -- osd_op(unknown.0.0:233 12.0 12:74038b3a:::rbd_header.10bda7b0bbc4:head [call rbd.migration_get in=16b] snapc 0=[] ondisk+read+known_if_redirected+supports_pool_eio e50) v8 -- 0x7fcd6009d5a0 con 0x195b050
2023-01-06T16:08:57.857+0000 7fcd3d7fa700  1 -- 10.89.0.28:0/968080585 <== osd.0 v2:10.89.0.26:6800/144 254 ==== osd_op_reply(233 rbd_header.10bda7b0bbc4 [call out=101b] v0'0 uv40 ondisk = 0) v8 ==== 167+0+101 (crc 0 0 0) 0x7fcd1403d020 con 0x195b050
2023-01-06T16:08:57.857+0000 7fcd6d586700 20 librbd::Migration: open_images: source migration spec: [header_type=source, pool_id=12, pool_namespace=, image_name=c91bb114-a4ae-436a-bb3f-0b2656d92650Í^?, image_id=10bd7cc71cf3, snap_seqs={}, overlap=0, flatten=0, mirroring=0, mirror_image_mode=journal, state=prepared, state_description=]
2023-01-06T16:08:57.857+0000 7fcd6d586700  5 librbd::Migration: status: migrating 7ea7e6f6-dafa-4597-878d-bbae856c446f/a52103d9-0c61-4280-96b1-bd065cdf4bf7 -> 7ea7e6f6-dafa-4597-878d-bbae856c446f/c91bb114-a4ae-436a-bb3f-0b2656d92650Í^?
```
...
```
2023-01-06T16:08:57.863+0000 7fcd077fe700  5 librbd::io::Dispatcher: 0x7fcd000051f0 shut_down: 
2023-01-06T16:08:57.863+0000 7fcd077fe700  5 librbd::io::ObjectDispatch: 0x7fcd6000c720 shut_down: 
2023-01-06T16:08:57.863+0000 7fcd077fe700 10 librbd::image::CloseRequest: 0x7fcd10057650 handle_shut_down_object_dispatcher: r=0
2023-01-06T16:08:57.863+0000 7fcd077fe700 10 librbd::image::CloseRequest: 0x7fcd10057650 send_flush_op_work_queue
2023-01-06T16:08:57.863+0000 7fcd07fff700 10 librbd::image::CloseRequest: 0x7fcd10057650 handle_flush_op_work_queue: r=0
2023-01-06T16:08:57.863+0000 7fcd07fff700 10 librbd::image::OpenRequest: handle_close_image: r=0
2023-01-06T16:08:57.863+0000 7fcd07fff700 10 librbd::ImageState: 0x7fcd10067200 0x7fcd10067200 handle_open: r=-2
2023-01-06T16:08:57.863+0000 7fccfeffd700 10 librbd::ImageCtx: 0x7fcd60046120 ~ImageCtx
2023-01-06T16:08:57.863+0000 7fccfeffd700 20 librbd::AsioEngine: 0x7fcd6000bb30 ~AsioEngine: 
2023-01-06T16:08:57.863+0000 7fccfeffd700 20 librbd::asio::ContextWQ: 0x7fcd60016240 ~ContextWQ: 
2023-01-06T16:08:57.863+0000 7fccfeffd700 20 librbd::asio::ContextWQ: 0x7fcd60016240 drain: 
2023-01-06T16:08:57.863+0000 7fcd6d586700 10 librbd::Migration: open_images: Source image is not found. Trying trash
2023-01-06T16:08:57.863+0000 7fcd6d586700 20 librbd::api::Trash: list: list 0x7fcd6d582800
2023-01-06T16:08:57.863+0000 7fcd6d586700 20 librbd::api::Trash: list_trash_image_specs: list_trash_image_specs 0x7fcd6d582800
2023-01-06T16:08:57.863+0000 7fcd6d586700  1 -- 10.89.0.28:0/968080585 --> [v2:10.89.0.26:6800/144,v1:10.89.0.26:6801/144] -- osd_op(unknown.0.0:238 12.0 12:c4c92e5a:::rbd_trash:head [call rbd.trash_list in=25b] snapc 0=[] ondisk+read+known_if_redirected+supports_pool_eio e50) v8 -- 0x7fcd6000e600 con 0x195b050
2023-01-06T16:08:57.863+0000 7fcd3d7fa700  1 -- 10.89.0.28:0/968080585 <== osd.0 v2:10.89.0.26:6800/144 259 ==== osd_op_reply(238 rbd_trash [call out=84b] v0'0 uv28 ondisk = 0) v8 ==== 153+0+84 (crc 0 0 0) 0x7fcd1403d020 con 0x195b050
2023-01-06T16:08:57.863+0000 7fcd6d586700 -1 librbd::Migration: open_images: failed to determine image id: (2) No such file or directory
    migration_test.go:151: 
        »»»»Error Trace:»»»»/go/src/github.com/ceph/go-ceph/rbd/migration_test.go:151
        »»»»Error:      »»»»Received unexpected error:
        »»»»            »»»»RBD image not found
        »»»»Test:       »»»»TestMigrationAbort
--- FAIL: TestMigrationAbort (6.12s)
FAIL
coverage: 7.0% of statements in github.com/ceph/go-ceph/rbd
FAIL»»»»github.com/ceph/go-ceph/rbd»20.244s
FAIL
*** ERROR: returned 1
*** ERROR: rbd tests failed
```

trociny commented:
@phlogistonjohn looking at the log
```
2023-01-06T15:45:45.103+0000 7efc21423cc0 10 librbd::Migration: status: d560a358-85a8-4096-b43f-898adbaa3763/f9c472a4-7840-4045-809b-cadb4c83ac89373~
2023-01-06T15:45:45.103+0000 7efc21423cc0 10 librbd::Migration: open_images: trying to open image by name d560a358-85a8-4096-b43f-898adbaa3763/f9c472a4-7840-4045-809b-cadb4c83ac89373~
2023-01-06T15:45:45.105+0000 7efc21423cc0 20 librbd::asio::ContextWQ: 0x20281c0 ContextWQ: 
2023-01-06T15:45:45.105+0000 7efc21423cc0 20 librbd::AsioEngine: 0x1f6c430 AsioEngine: 
2023-01-06T15:45:45.105+0000 7efc21423cc0 10 librbd::ImageCtx: 0x1f32a70 ImageCtx: image_name=f9c472a4-7840-4045-809b-cadb4c83ac89373~, image_id=
```
That `~` at the end of the image name (`d560a358-85a8-4096-b43f-898adbaa3763/f9c472a4-7840-4045-809b-cadb4c83ac89373~`) looks suspicious. I suspect when you copy the image name from the go to the c you don't provide the terminating `'\0'`.

phlogistonjohn commented:
Great. That's very helpful indeed! 

@omidyoosefi you are calling CBytes instead of CString in most of these functions. I missed this because otherwise it looks like the same pattern we typically use when converting between Go and C strings with CString.  CBytes is typically used with C functions that take a buffer (byte array) and length. Where CString is more properly used for Null-terminated strings.

@trociny thanks for the assistance. I wish I had noticed this sooner, and hadn't bothered you about it, but I appreciate the help.

omidyoosefi commented:
@phlogistonjohn @trociny

Thank you for pursuing this while I was away on vacation. So the issue was in the implementation, UTs for the win! I have made the change and tried another 100 tests all passing.

trociny commented on rbd/migration_test.go:
```diff
@@ -0,0 +1,164 @@
+	require.Eventually(t, func() bool {
+		status, err = MigrationStatus(ioctx, destImage)
+		return err == nil
+	}, 5*time.Second, 10*time.Millisecond)
```
I don't think you need this `require.Eventually` thing. It should succeed on the first call. And the same for all other cases below.

phlogistonjohn replied:
Agreed, please try removing these and if the runs become unstable without them we'll just need ton investigate further.

omidyoosefi replied:
Made the change.

phlogistonjohn commented on rbd/migration.go:
```diff
@@ -0,0 +1,219 @@
+// Implements:
+//  int rbd_migration_prepare(rados_ioctx_t ioctx,
+//                            const char *image_name,
+// 	                          rados_ioctx_t dest_ioctx,
```
There looks to be some differing whitespace on this line.

omidyoosefi replied:
fixed

phlogistonjohn commented on rbd/migration.go:
```diff
@@ -0,0 +1,219 @@
+//                            const char *dest_image_name,
+//                            rbd_image_options_t opts);
+//
+func MigrationPrepare(ioctx *rados.IOContext, destIoctx *rados.IOContext, sourceImageName string, destImageName string, rio *ImageOptions) error {
```
Let's match that order of arguments in the c-call. So `ctx, name, ctx, name, ...` instead of `ctx, ctx, name, name`.  We usually try to stick to matching the underlying C apis for things like this.

omidyoosefi replied:
done

phlogistonjohn commented on rbd/migration.go:
```diff
@@ -0,0 +1,219 @@
+//                                   const char *dest_image_name,
+//	                                 rbd_image_options_t opts);
+//
+func MigrationPrepareImport(ioctx *rados.IOContext, sourceSpec string, destImageName string, rio *ImageOptions) error {
```
See above. Let's match the order of args in the C apis.

omidyoosefi replied:
done

phlogistonjohn commented on rbd/migration.go:
```diff
@@ -0,0 +1,219 @@
+// from a specified source to a new target image
+//
+// Implements:
+//  int rbd_migration_prepare_import(const char *source_spec,
+//	                                 rados_ioctx_t dest_ioctx,
+//                                   const char *dest_image_name,
+//	                                 rbd_image_options_t opts);
```
Differing whitespace on some of these lines.

omidyoosefi replied:
fixed

phlogistonjohn reviewed (changes requested):
OK, I have done another pass through the patches now that we've fixed the major issue. I found a couple indent issues in the doc comments. Plus, could you please end the first sentances of the doc comments with periods. That will match the style used elsehwere for Go doc comments.

Finally, I noticed that a couple of the functions have differing argument orders from the C apis. Generally, we try to match the APIs in C. This is particularly useful if people are familiar with the API in one language already :-)

phlogistonjohn commented:
The ceph `main` test job is failing due to what I think are issues with the ceph infra. The rest of the jobs are detecting the new API has not yet been added to the api status JSON (and related) files.  It should just be a matter of running    `make api-update` and committing the updated files.


For context:  [API Status in the Developers Guide](https://github.com/ceph/go-ceph/blob/master/docs/development.md#api-status) and the [API Stability Plan](https://github.com/ceph/go-ceph/blob/master/docs/api-stability.md).

I've filed a PR to  update the pull request template so we don't forget to mention  these as often as we do.

omidyoosefi commented:
> +1 for matching C api as close as possible, unless it violates your go standards.
> 
> Also is there any special reason you named `rbd_image_migration_status_t` struct as `MigrationImageInfo` (not `MigrationImageStatus`)? For me it would be confusing, because when I see `ImageInfo` I think about an image (quasi-static) properties, while the status is more about the state (of the migration in this case).

No special reasoning, made the change.



> The ceph `main` test job is failing due to what I think are issues with the ceph infra. The rest of the jobs are detecting the new API has not yet been added to the api status JSON (and related) files. It should just be a matter of running `make api-update` and committing the updated files.
> 
> For context: [API Status in the Developers Guide](https://github.com/ceph/go-ceph/blob/master/docs/development.md#api-status) and the [API Stability Plan](https://github.com/ceph/go-ceph/blob/master/docs/api-stability.md).
> 
> I've filed a PR to update the pull request template so we don't forget to mention these as often as we do.

Ah, yes. I ran the make command and added the changes.

anoopcs9 commented on docs/api-status.md:
```diff
@@ -12,10 +12,19 @@ No Preview/Deprecated APIs found. All APIs are considered stable.
 
 ## Package: rados
 
-No Preview/Deprecated APIs found. All APIs are considered stable.
-
```
You can take out this change(unrelated) as it might be a bug with the `make api-update` process.

omidyoosefi replied:
Thank you for the review @anoopcs9. Made the change

anoopcs9 commented on docs/api-status.md:
```diff
@@ -30,8 +39,6 @@ No Preview/Deprecated APIs found. All APIs are considered stable.
 ## Package: rgw/admin
 
-No Preview/Deprecated APIs found. All APIs are considered stable.
-
```
Same comment as above.

omidyoosefi replied:
Done

phlogistonjohn commented:
Almost there! I just noticed one more small "bureaucratic" item: we generally prefer commits contain a Signed-off-by/DCO line.  I think it should just be a matter of running `git commit --am -s`.

phlogistonjohn commented:
Also, `main` CI job continues to fail due to Ceph infra issues, but we do not need it to pass in order to merge IIRC.

omidyoosefi commented:
> Almost there! I just noticed one more small "bureaucratic" item: we generally prefer commits contain a Signed-off-by/DCO line. I think it should just be a matter of running `git commit --am -s`.

Done!

phlogistonjohn reviewed (approved):
Thank you very much for your contribution, LGTM.

phlogistonjohn commented:
@anoopcs9 can you please re-approve as your earlier review was dismissed by the force push to add the sign-off

anoopcs9 reviewed (approved):
Thank you.

omidyoosefi commented:
Thank you all for your support and help in getting this API in.
